ENH: support float and longdouble in FFT using C++ pocketfft version

mhvk opened the pull request:
Using the C++ version of pocketfft turned out to be not that much of a hassle. With it, we can support all the `dtype`s that it supports: float, double, and long double.

@asmeurer, @mreineck - comments are most welcome! (My C++ is non-existent so this really is just a C file with templates thrown in, mostly using trial & error...)

fixes #17801

EDIT: some things to decide:
- [x] How should we add the license? Right now just added `LICENSES.md` in the `pocketfft` directory. (The licence wasn't there for the C version, but that seems to have been an oversight.)
- [x] Does it make most sense to include pocketfft as a git submodule? This would include the license (and a small demo programme)
- [X] Should we use the cache? -> No (@mreineck suggested not to given the extra memory usage; scipy does use 16)
- [X] Should we enable threading? (scipy exposes it but uses 1 thread by default, see its [meson.build](https://github.com/scipy/scipy/blob/main/scipy/fft/_pocketfft/meson.build)) -> leave for later (A quick check on my laptop shows it gives about a rather array-size dependent speed-up of about a factor 2 even when using all 12 cores, so likely memory bound)
-  [x] Pocketfft will just bail if it cannot allocate memory (for its twiddle factors, and I guess a possible cache). ~If that is something we should catch, please advice how to intercept `throw`...~ dealt with using `try/catch`.
- [x] More tests? (push to this PR welcome.)

mhvk committed:
MAINT: Replace pocketfft with the C++ version and use it in ufuncs

Note that the C++ version is header only.  We just link to upstream
via a git submodule.

mhvk committed:
ENH: Use templating to add support float and longdouble.

Also adjust python side to allow multiple double types.

mhvk commented on numpy/fft/_pocketfft_umath.cpp:
```diff
@@ -362,8 +379,14 @@ add_gufuncs(PyObject *dictionary) {
 
 static struct PyModuleDef moduledef = {
```
Windows C was happy with the old version, but windows C++ is not...
```
../numpy/fft/_pocketfft_umath.cpp(382): error C7555: use of designated initializers requires at least '/std:c++20'
```

seberg replied:
Right, a bit unfortunate, but not surprising.

mreineck commented:
Ah, OK, this switches to the C++ version, but uses only its 1D transforms!
Makes a lot of sense for a quick fix of the data type issue, but it misses the chance of massive speedups (through vectorization and/or multithreading) in multi-D transforms.
If `numpy 2.0` is close that's probably the best option, though I would still urge to unlock the other features in later releases.
If only 1D FFTs are used, a lot of C++ code could be `#if 0`'d to reduce compile time. I can do this once the rest of the PR has converged.

seberg commented:
> but it misses the chance of massive speedups (through vectorization and/or multithreading) in multi-D transforms.

That is actually a (small or not?) issue with using the gufuncs, it may be tricky to expand to arbitrary dimensions in a nice way for those (which was mentioned to me, I admit; although I didn't realize it might be an issue any time soon).

And yes, so long as this looks all fine, it seems fine to not do the N-D versions.

mhvk commented:
Indeed, I went the easy route, based on the gufunc I had just introduced (mostly to be able to allow `out`). Is the threading/vectorization specific to the multi-dimensional iterator, or something that makes the 1D transform faster? From a quick look, it looks like it is the latter, so that should be relatively straightforward, by just using `c2c`, etc., in my calls. Indeed, that would mean no need for creating a buffer any more too...

mreineck commented:
> Is the threading/vectorization specific to the multi-dimensional iterator, or something that makes the 1D transform faster?

No, threading/vectorization only affects 2D and higher transforms. I have played with applying this to 1D as well, but only in `pocketfft`'s successor package.

seberg commented on numpy/fft/_pocketfft_umath.cpp:
```diff
@@ -20,41 +20,47 @@
 static inline void
 copy_data(char* in, npy_intp step_in, npy_intp nin,
-          char* out, npy_intp step_out, npy_intp nout, npy_intp elsize)
+          char* out, npy_intp step_out, npy_intp nout)
```
Two comments:
1. Let's use `T *out` and `T *in` here and then follow up with using `*out = *in` below and also `*out = 0`.  (I.e. remove all memcpy/memset, it was just annoying without templating.) 
2. Don't use that `pocketfft:detail::cmpl` thingy, just use `std::complex` (with `#include <complex>`.  However, I am not sure if there be dragons with migw longdouble:  You may have to use the LongDouble defined in `_core/src/npstd.hpp` (that pulls in complex also).

mhvk replied:
> 1. Let's use `T *out` and `T *in` here and then follow up with using `*out = *in` below and also `*out = 0`.  (I.e. remove all memcpy/memset, it was just annoying without templating.)

And adjust the step sizes accordingly in the call, I assume?

mhvk replied:
> 2\. Don't use that `pocketfft:detail::cmpl` thingy, just use `std::complex` (with `#include <complex>`.  However, I am not sure if there be dragons with migw longdouble:  You may have to use the LongDouble defined in `_core/src/npstd.hpp` (that pulls in complex also).

I think I need `cmplx` to be able to pass it in to the pocketfft functions without getting warnings (this likely mostly because I'm using internal routines).

seberg replied:
Ah, sorry, I forgot about striding...

You can't do that, since step sizes might be half-way or so, we could assign via `*(T *)out = 0`, which is still a bit nicer I would say?

seberg replied:
OK, I mainly thought to replace it to not use it too much, and I thought that assignments seem to maybe not work for that one, but maybe they do anyway...

If this seems no improvement, then don't worry too much, that seemed to me like it might with a brief look, but not sure.

mhvk replied:
@seberg - Do you mean something like,
```
        if (step_in == 1 && step_out == 1)) {
            for (npy_intp i = 0; i < ncopy; i++) {
                out[i] = in[i];
            }
        }
        else {
            T *ip = in, *op = out;
            for (npy_intp i = 0; i < ncopy; i++, ip += step_in, op += step_out) {
                *op = *ip;  // Or without pointers? out[op] = in[ip]
            }
        }
```
Or just a single loop? (but then it is less obvious whether the compiler can still optimize the contiguous case)

mhvk replied:
I tried and liked it at some level, but got stuck at being able to set `out[i] = 0` which `cmplx` type does not know how to interpret. So maybe indeed best to just keep things as they are...

seberg replied:
Right, I thought that would the case and that using `std::complex` would allow it.  Have to look at the rest, but for now let's just keep it as is, I guess.

mhvk commented:
Just to be clear, do you mean that vectorization helps for FTs over multiple axes, or FTs over a single axis in higher-D arrays? From the code, it looks like the latter, which might still work with the ufunc, as we get an extra axis passed in.

> No, threading/vectorization only affects 2D and higher transforms. I have played with applying this to 1D as well, but only in `pocketfft`'s successor package.

OK, that seems reasonable for follow-up, then. More generally, it may well make sense to use your iterator too, and/or copy what scipy does... I definitely went for what seemed the fastest route...

asmeurer commented on doc/release/upcoming_changes/25536.new_feature.rst:
```diff
@@ -1,4 +1,5 @@
-The various FFT routines in `numpy.fft` have gained an ``out``
+`numpy.fft` support for different precisions and in-place calculations
+----------------------------------------------------------------------
+The various FFT routines in `numpy.fft` now support calculations in float,
```
They have always supported single precision but have upcast the result. We should be clear that this is a breaking change.

mhvk replied:
Does this sound good?
```
The various FFT routines in `numpy.fft` now do their calculations natively in
float, double, or long double precision, depending on the input precision,
instead of always calculating in double precision. Hence, the calculation will
now be less precise for single and more precise for long double precision.
The data type of the output array will now be adjusted accordingly.

Furthermore, all FFT routines have gained an ``out`` argument that can be used
for in-place calculations.
```

seberg replied:
Maybe it makes sense to have a dedicated change note that points out the single precision change explicitly.

asmeurer replied:
What's the numpy policy on using [`versionchanged`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionchanged). Should that be added to the fft docstrings?

asmeurer replied:
> Maybe it makes sense to have a dedicated change note that points out the single precision change explicitly.

I don't know how the numpy release notes are amalgamated, but for 2.0 it would be good to highlight all the compatibility breaks somewhere in the release notes.

mreineck commented:
> Just to be clear, do you mean that vectorization helps for FTs over multiple axes, or FTs over a single axis in higher-D arrays? From the code, it looks like the latter, which might still work with the ufunc, as we get an extra axis passed in.

Yes, it works for transforming just a single axis in a multi-D array too. The idea is to "bundle" a few of the 1D transforms into a single one of SIMD type.

mhvk commented:
The `macos_arm64_test` failure seems unrelated...

mhvk commented:
@mreineck - OK, I'll look into using `c2c` and friends...

mhvk commented:
Calling `c2c`, etc., using the outer dimension actually seems fairly straightforward and gives a nice speed-up. The only unfortunate thing is that pocketfft does not deal with the case that the input is shorter than the output, so in that case one still needs to buffer/copy. But the speed-up is worth the complexity, so I'll push a further refactoring tomorrow.

rgommers commented:
Nice!

> Furthermore, I have not tried setting up a pocketfft cache, or done anything with threading. Scipy uses its [meson.build](https://github.com/scipy/scipy/blob/main/scipy/fft/_pocketfft/meson.build) to set a cache size of 16 and have compiler-dependent threading; happy to do the same here.

The cache size may help, assuming it's a useful setting in single-threaded mode (@mreineck?). I'd leave the threading along, we can't do much with that. In SciPy it's controlled by a `workers=` keyword to opt in to multi-threading. NumPy doesn't have that, and everything is supposed to be single-threaded (underlying BLAS/LAPACK is an exception for [reasons]).
 
> Should the license be added to LICENSES.txt? It wasn't there for the C version, but perhaps that was an oversight.

Probably an oversight indeed, since the SciPy licenses list has it ([here](https://github.com/scipy/scipy/blob/96871841bdbb121729f63837709e725a6e285b7f/LICENSES_bundled.txt#L207-L210)). So +1 for adding it.

> With it, we can support all the `dtype`s that it supports: float, double, and long double.

Adding `float32` support is great and obviously desirable. I'm less sure about `longdouble` support, since that's not really all that useful and something we may have more work to get rid of again later. We didn't get around to revisiting this topic for 2.0, but overall I'm still convinced that `longdouble` is overall more harmful than helpful for NumPy as a whole, so why add more support when I don't think anyone asked for it here?

Build time wise this PR in its current state looks fine - impact of C++ is okay overall, see quick profiling result below. Binary size is fine too, it's ~240 kb for a dev (debugoptimized) build. A note though that the SciPy pocketfft extension is more than 4x larger, so enabling all functionality that SciPy uses in the NumPy build is probably not the way to go (it'd be ~25% of the size of `_multiarray_umath` at that point).

<img width="1210" alt="image" src="https://github.com/numpy/numpy/assets/98330/026b633f-d530-486a-9d0d-a5f306b51b71">

mreineck commented:
> The cache size may help, assuming it's a useful setting in single-threaded mode (@mreineck?)

It will work in single-threaded mode. However, my feeling is that the plan caching isn't really worth it: it may cause large memory consumption (see, e.g., https://github.com/scipy/scipy/issues/12297) for very little performance gain (unless you call very short 1D FFTs repeatedly). `Scipy` maintainers tend towards keeping the cache, however.

mhvk committed:
MAINT: use vectorized pocketfft routines when possible

Also make sure every part uses copy_input and copy_output
consistently, and avoid compiling multithreading.

mhvk commented:
> Adding `float32` support is great and obviously desirable. I'm less sure about `longdouble` support, since that's not really all that useful and something we may have more work to get rid of again later.

@rgommers - I would love to have a proper `float128` instead of `longdouble`, but on the other hand, the present behaviour is not that `longdouble` is not supported (in which case it would be fine to continue with that), but rather that `longdouble` is just converted to `double` without any warning. Given that, my sense is that supporting it correctly is the least disruptive change.

rgommers commented:
> the present behaviour is not that `longdouble` is not supported (in which case it would be fine to continue with that), but rather that `longdouble` is just converted to `double` without any warning. Given that, my sense is that supporting it correctly is the least disruptive change.

I'm not really opposed, but it is a change, with technically only increased binary size as a penalty and no upsides for Windows and macOS arm64 users. Doing nothing and continuing to convert to `float64` either internally like before or, as an improvement, use the float64 loops internally but cast back to `longdouble` before returning the result, are therefore also options (even the former is zero change from now, so hard to label it as more disruptive). 

I'm not sure how serious the plan was to include all the pocketfft functionality that SciPy needs into NumPy - but binary size does seem to become relevant then. If's it's <0.1 MB I'm not too worried, but I wouldn't really want to pay 0.25 MB or more for long double loops that no one has asked for.

mhvk commented:
```
ls -lh build/numpy/fft/_pocketfft_umath.cpython-311-x86_64-linux-gnu.so
# 5.1M present PR
# 4.1M without longdouble
# 2.9M w/ longdouble but with POCKETFFT_NO_VECTORS
# (which uses low-level routines only, but halves speed;
# I checked that high-level routines are not compiled for longdouble)
```
For comparison, `multiarray_umath` is 73 MB, the simd stuff 15MB, and everything in random adds up to 43 MB.

rgommers commented:
Thanks for the comparison, that is helpful. Note that those are `debugoptimized` numbers; what I had in mind was wheel sizes (I should have specified). Our wheel sizes are in the 13 - 21 MB range, and the new macOS arm64 ones even as small as 5 MB.

I did a quick test building a wheel from this branch; the pocketfft extension is 508 kb uncompressed, 208 kb compressed in it (total wheel size 7.8 MB; `_multiarray_umath.so` 2.6 MB compressed). So that should still be fine. If dst/dct transforms would be enabled, it may get a bit questionable. But that could still be taken care of mostly by disabling the `longdouble` loops for platforms where it's an alias for `double`. Anyway, it doesn't seem too likely to become an issue at this moment. Let's not worry about it further.

mhvk commented:
OK, great! The other transforms may not be as bad as one would think, since they mostly use the same code. But anyway, not sure those are needed, and definitely not relevant here.

mreineck commented on numpy/fft/_pocketfft_umath.cpp:
```diff
@@ -0,0 +1,434 @@
+#include "numpy/ufuncobject.h"
+
+#include "npy_config.h"
+
```
You may want to add an

`#define POCKETFFT_NO_MULTITHREADING`

here. Not sure if it saves much space, but since `numpy` is not going to use parallelism anyway, it cannot hurt.

mhvk replied:
Yes, nice! On the debug build, it reduces the size from 5.1 to 4.7 MB, so helpful.

mhvk commented:
OK, I think this is basically ready now, though there are a few more general questions - see the list on top.

mreineck commented:
Concerning points 1 and 2 of the list, it's probably useful to include `pocketfft` as a submodule. `scipy` did the same recently and introduced a fork at `https://github.com/scipy/pocketfft` for this purpose.

Concerning error handling, it would be great to have wrapper code which translates C++ exceptions into Python ones. `pybind11` (used for wrapping `pocketfft` in `scipy`) does that for us, but I don't know how hard it would be to do something analogous here. It would come in handy for any future C++ components of `numpy` too.

seberg commented:
I will probably try to get started on my own bigger PRs before looking into this, but nice to see this on the way to get into 2.0.  (will manage some time next week to review for sure, but if someone is faster, that is better)

It would be bad not to intercept the C++ errors, I am not sure how easy that is.  We are setting some compiler flags to *not* throw exceptions ourselves, but hopefully catching them still works(?):
```
try {
}
catch (...) {
}
```
We should catch the memory error specifically, I think.  And everything else translate to some generic `RunetimeError`.

I don't think it is plausible to let the error bubble higher than the ufunc loo right now.  That would cross the C/C++ boundary.

mreineck commented:
> It would be bad not to intercept the C++ errors, I am not sure how easy that is. We are setting some compiler flags to not throw exceptions ourselves [...]

Ouch ... that's probably fatal in this context. `pocketfft` relies on standard C++ behavior, e.g. `new` throwing an exception when memory cannot be allocated. If this is disabled via compiler flags and the `new` call just returns NULL instead, there will be segmentation faults.

seberg commented:
OK, not sure how broadly flags are set, but in most places we need to ensure we don't accidentally bubble up exceptions into C, thus they are set.  So we should double check this unfortunately.
I would think we can locally compile with exceptions here and just make very sure to catch them.  @seiko2plus maybe you can give a hand with this?

mhvk commented on numpy/fft/_pocketfft_umath.cpp:
```diff
@@ -0,0 +1,437 @@
+     * Otherwise, use a non-vectorized loop in which we try to minimize copies.
+     * We do still need a buffer if the output is not contiguous.
+     */
+    auto plan = pocketfft::detail::get_plan<pocketfft::detail::pocketfft_c<T>>(nout);
```
@mreineck - for guarding against out-of-memory, getting the plan here would be relevant, but not the actual execution, correct?

Plus of course the call to the iterator `pocketfft:c2c` above.

These two get repeated twice more, but it should be easy to abstract that a bit if needed.

mhvk commented:
These are the compilation flags I see in my `spin` call (from `build.ninja`):
```
-Wall -Winvalid-pch -std=c++17 -O2 -g -msse -msse2 -msse3 -DNPY_HAVE_SSE2 -DNPY_HAVE_SSE -DNPY_HAVE_SSE3 -fPIC
```
I don't see anything that changes c++ behaviour.

seiko2plus commented:
> The relevant flags are here: 

Its only limited to core only, I thought it was a global flag

mhvk commented:
Are we sure those flags are seen in `fft` -- since they don't appear in my build log... I'm not sure how `meson` exactly works, but does what happens in `subdir(_core)` influence `subdir(fft)? See 
https://github.com/numpy/numpy/blob/2634f803313f349170c09606d3cc619accd72247/numpy/meson.build#L398-L401

mhvk commented:
OK, I just empirically confirmed that that flag does not get passed on. And with that, that I can catch C++ exceptions. I'll push that up in a bit.

seiko2plus commented:
> And with that, that I can catch C++ exceptions

If its about memory allocations, I think we should use our custom allocations over malloc or new.

mhvk committed:
MAINT: handle possible c++ errors by wrapping the ufuncs

Also use internal pocketfft arr class to allocate the possible
temporary buffers, so we do not have to worry about deallocating
them any more.

mhvk commented:
@seiko2plus - the problem is that the allocations (and possible other errors) happen inside `pocketfft` and ideally we do not start rewriting that...

My current solution is just to put everything in a `try/catch` using macros and convert C++ exceptions to python errors. I don't really know how to test memory errors, but I checked that if I purposely pass in an invalid axis number, the corresponding exception properly leads to a `RuntimeError`. See last commit, ab16ef41a973bd00553af2c91953412a66d8c365

mreineck commented:
> If its about memory allocations, I think we should use our custom allocations over malloc or new.

I'd really recommend not doing that. Most likely these custom operations are C-style and therefore will silently return NULL on error, which goes against fundamental assumptions C++ codes are making. You'd have to patch every place in pocketfft where memory is allocated and add error handling code.

mhvk commented:
p.s. Note that I now on purpose use `pocketfft` allocators, since those throw an exception, hence I could remove my custom code.

mreineck commented:
> My current solution is just to put everything in a try/catch using macros and convert C++ exceptions to python errors.

To me this looks exactly like the right solution. It's located directly at the interface, does not require any big patching and can be extended in case more detailed error handling is required.
If memory allocations behave in the standard C++ way, they will be caught here as well.

mreineck commented:
`"-fno-exceptions"` should certainly not be used when compiling `pocketfft` and the wrapper code.
To be honest, I'd consider removing `-fno-exceptions -fno-rtti` from your configuration altogether. This was a good idea 20 years ago, but today the overhead (both memory and runtime) should be pretty negligible.

seberg commented on numpy/fft/_pocketfft_umath.cpp:
```diff
@@ -0,0 +1,437 @@
+    void *buff = NULL;
+    if (step_out != sizeof(std::complex<T>)) {
+        buff = pocketfft::detail::aligned_alloc(sizeof(std::complex<T>),
+                                                nout * sizeof(std::complex<T>));
```
There is a small problem here that `buff` is a `void *` and not a `pocketfft:detail::arr<std::complex<T>>`.  Which I think means that if the `plan->exec` fails below no cleanup will occur.
My C++ is a bit trial and error, but I think:
```
/* Create a temporary array to hold a buffer if op is not contiguous */
bool contiguous_op = (step_out != sizeof(std::complex<T>));
pocketfft::detail::arr<std::complex<T>> tmp_arr(contiguous_op ? 0 : nout);
```
should be an option (an array of length 0 doesn't do a malloc so that is fine).

And below:
```
std::complex<T> *op_or_buff = (contiguous_op ? (std::complex<T> *)op : tmp_arr.data());
```

(Clearly, the dealloc below then becomes unnecessary, which is the whole point.)

mhvk replied:
OK, that makes a lot of sense! I implemented it (though with an opposite bool, `buffered`).

seberg commented on numpy/fft/_pocketfft_umath.cpp:
```diff
@@ -0,0 +1,437 @@
+        PyErr_SetString(PyExc_RuntimeError, e.what());
+    }
+    NPY_DISABLE_C_API;
+}
```
I don't really love the way it puts a bracket in the macro (yes other places do too).  This would be an alternative.  Haven't made up my mind if I much prefer it.  I prefer the pattern, but it adds churn further down.
```
/*
 * In order to ensure that C++ exceptions are converted to Python
 * ones before crossing over to the C machinery, we must catch them.
 * This template can be used to wrap a C++ written ufunc to do this via:
 *      wrap_legacy_cpp_ufunc<cpp_ufunc>
 */
template<PyUFuncGenericFunction cpp_ufunc>
static void
wrap_legacy_cpp_ufunc(<full ufunc signature>)
{
    NPY_ALLOW_C_API_DEF
    try {
        cpp_ufunc(<full ufunc signature>)
    }
    catch (std::bad_alloc& e) {
        NPY_ALLOW_C_API;
        PyErr_NoMemory();
        NPY_DISABLE_C_API;
    }
    catch (const std::exception& e) {
        NPY_ALLOW_C_API;
        PyErr_SetString(PyExc_RuntimeError, e.what());
        NPY_DISABLE_C_API;
    }
}
```
Then use `wrap_legacy_cpp_ufunc<irfft_loop<double>>` below.

mhvk replied:
That's a really nice idea -- implemented!

seberg reviewed (commented):
Not a deep review (although I doubt that is needed at this point!).
Just commenting on two things so they don't get lost (one bug, one suggestion to try to have a more C++ pattern for catching the exceptions).

Submodule+license seem still two things that would be nice to do as well.

mhvk commented:
@seberg - thanks! Both suggestions were really nice, and make the code a lot cleaner. I pushed a revised version that uses them and that also replaces the addition of the C++ pocketfft header with a new submodule (which includes the license).

mhvk commented:
p.s. In the checklist above, this just leaves the question of whether this is tested thoroughly enough. I could adjust more tests to try different `dtype`, but I'm a bit wary of just increasing test run time for little benefit. Currently, the tests that I did change exercise the options to do longer or shorter FTs than the input (i.e., which set `n`), which has been the main pain point so far.

mhvk commented:
p.p.s. Well the above about tests was not actually true, but now it is, so I ticked my "more tests" item...

mhvk commented on numpy/fft/tests/test_pocketfft.py:
```diff
@@ -381,15 +389,6 @@ def test_all_1d_norm_preserving(self):
                     assert_allclose(x_norm,
                                     np.linalg.norm(tmp), atol=1e-6)
 
-    @pytest.mark.parametrize("dtype", [np.half, np.single, np.double,
```
This one is now done much more thoroughly in `test_identity_long_short` and `test_identity_long_short_reversed`

mhvk commented:
Is there anything more to do for this? (C++ pocketfft in `np.fft`)

seberg reviewed (approved):
Not managing to concentrate very deeply on it, but those two C++ things were basically what I noticed and the rest looks fine.

Since nobody voiced other concerns, let's get this in.  (It isn't vastly changed after all, since it is porting the C code.)


Thanks for this great new code!
